pkg/pma/wal.c

```diff
@@ -0,0 +1,350 @@
+/// @file
+
+#include "wal.h"
+
+#include <assert.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "murmur3.h"
+#include "util.h"
+
+//==============================================================================
+// CONSTANTS
+
+/// Seed value for hashing function used to compute page checksum.
+static const size_t kSeed = 0;
+
+//==============================================================================
+// TYPES
+
+/// An entry in a WAL's metadata file.
+struct metadata_entry_ {
+    /// Page index.
+    uint64_t pg_idx;
+    /// Checksum of page contents.
+    uint64_t checksum;
+};
+typedef struct metadata_entry_ metadata_entry_t_;
+
+//==============================================================================
+// STATIC FUNCTIONS
+
+/// Open a file.
+///
+/// @param[in]  base_path  Path to base file.
+/// @param[in]  suffix     Suffix to append to base file path.
+/// @param[out] path       Populated with the concatenation of base_path and
+///                        suffix.
+/// @param[out] fd         Populated with an open file descriptor for file at
+///                        path.
+/// @param[out] len        Populated with length of file at path.
+///
+/// @return 0   Success.
+/// @return -1  Failure.
+static int
+open_file_(const char *base_path,
+           const char *suffix,
+           char      **path,
+           int        *fd,
+           size_t     *len);
+
+static int
+open_file_(const char *base_path,
+           const char *suffix,
+           char      **path,
+           int        *fd,
+           size_t     *len)
+{
```

Comment:
It's actually not. I find it's useful to declare and define a static function separately so any static function can call any other static function regardless of the order they're defined in (since all static functions are declared before any static functions are defined). The declaration also provides a logical place to document the static function. It's overkill when there's only one static function like `wal.c`, but I decided to stick with the pattern so it's easier to adhere to the pattern if another static function needs to be added in the future.